docs/Rest-API.md

```diff
@@ -0,0 +1,147 @@
+# Axe REST API
+
+Documentation of the REST API that comes along with the Axe service.
+
+## AxeResult
+
+This endpoint serves as a wrapper to collect all of the information from an axe run. 
+
+### Motivation
+
+- Provide access to results and context easily.
+- Avoid timing issues that come with multiple requests.
+
+### Simplified Sample
+```
+{
+  "axeContext" : { @AxeContext },
+  "axeResult" : { AXRL },
+  "axeConf" : { TBD by Framework }  
+}
+```
+
+### Requirements
+
+- Data points are collected at the same point in time, or as close as can be expected.
+
+## AxeContext
+
+Wraps data together that is needed to identify a particular View, with a particular screenshot,
+for a particular device configuration.
+
+### Motivation
+
+- Collect View information in one place.
+- Allow for direct correlation of screenshot and ViewHierarchy
+- Easy developer debugging.
+- Reproducability of results.
+
+### Simplified Sample
+
+```
+{
+  "axeView" : @AxeView,
+  "axeDevice" : @AxeDevice,
+  "axeEvents" : @AxeEvents,
+  "screenshot" : "base64png data"
+}
+```
+
+## AxeView
+
+Serves up properties about the Accessibility Hierarchy of a particular view. 
+
+### Motivation
+
+- All motivations of AxeContext apply.
+- Rule debugging.
+- Simplified test case collection.
+- Simplified false positive identification.
+
+### Simplified Sample
+
+```
+{
+  "Arbitrary Property Name": "value1",
+  "Arbitrary Property Name2": "value2",
+  "children" : [
+    {
+      "Arbitrary Property Name": "value3",
+      "Arbitrary Property Name2": "value24"
+      "children": []
+    }, 
+    {
+      "Arbitrary Property Name": "value3",
+      "Arbitrary Property Name2": "value24",
+      "children" : []
+    }
+  ]
+}
+```
+
+### Requirements
+
+- The Arbitrary property names of the children are identical to the root. 
+- If a property is omitted from a child it is assumed to be present and NULL.
+- Arbitrary property names are NOT calculated, but are values present in the respective OS APIs.
+
+## AxeDevice
+
+Collect prescriptive information that helps identify what type of context various information 
+was collected from.
+
+### Motivation
+
+- Identify device version and operating system.
+- Provide metrics for screen size so view hierarchy can be correlated.
+
+### Simplified Sample
+
+```
+{
```

Comment:
Should/can we include the device model? Sometimes the same OS can behave differently on different devices.

Reply:
This REST API is purposely designed to be very flexible. Particularly between iOS and Android. Any time a property is listed specifically it is required to be exactly that. Adding "model" to this would potentially create issues for iOS. 

What would you think about going with a non-structured "description" approach, which would include this information in a non-structured way for compatibility and clarity between the two frameworks.

Reply:
Trying to parse meaningful data from non-structured data is something that I'd strongly suggest that we avoid because it's inherently brittle. In my experience, you can safely do only 2 things with non-structured fields:
1. Display them in the UI
2. Group by them in reports/telemetry

If we want the client to offer greater functionality (such as programmatically identifying this as a Samsung devices), then a dedicated field--even an optional one--is far more robust.

---
In reply to: [279103399](https://github.com/dequelabs/axe-android/pull/11#discussion_r279103399) [](ancestors = 279103399)